### PR TITLE
[XLA:GPU] Default to flash attention and remove mask input in cuDNN

### PR DESCRIPTION
**Pull request by Cjkkkk**
* Default to flash attention as it is high performant and maintained actively by cuDNN, remove old fused attn.
  * Remove lowering to fused attn in rewriter
  * Remove cudnn graph generation
* Remove mask input as it is only doing multiply instead of masking with -inf, give incorrect results. Also cuDNN does not support this anymore, mask should be combined with bias. This is follow up on https://github.com/openxla/xla/pull/11444.
  * Remove mask logic in rewriter
  * Remove mask buffer/descriptor in thunk
* Remove bmm1-bmm2 pattern as it is not support by flash attention. Modified related rewriter test to use bmm1-softmax - bmm2.

Current pattern:
bmm1 - (scale) - (bias) - softmax - (dropout) - bmm2.

### CODE REVIEW

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 1]
```diff
@@ -1089,8 +1078,7 @@ absl::Status IrEmitterUnnested::EmitFusedMHABackwardThunk(
     mask_shape = instr->operand(input_index++)->shape();
   }
 
-  bool has_bias = is_flash_attention &&
-                  (kind == CudnnfMHAKind::kBackwardScaleBiasSoftmax ||
+  bool has_bias = (kind == CudnnfMHAKind::kBackwardScaleBiasSoftmax ||
                    kind == CudnnfMHAKind::kBackwardScaleBiasSoftmaxDropout ||
                    kind == CudnnfMHAKind::kBackwardScaleBiasMaskSoftmax ||
```
Comment (Tixxx) [inline xla/service/gpu/ir_emitter_unnested.cc:1083]: do we still need to keep those CudnnfMHAKinds with mask?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 2]
```diff
@@ -1650,34 +1420,27 @@ absl::StatusOr<bool> FuseBwdMultiHeadedAttentionBlock(
...
-      operands.push_back(bias);
-    }
-    HloInstruction* fwd_output;
-    for (auto user : fwd_fmha_call->users()) {
-      if (user->opcode() == HloOpcode::kGetTupleElement &&
-          user->tuple_index() == 0) {
-        fwd_output = user;
-      }
+  // For flash attention, add fwd output to input list
+  if (!is_causal_mask && bias) {
```
Comment (Tixxx) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:1429]: If mask is always combined with bias, should we always add bias here?

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 3]
```diff
@@ -1609,24 +1389,14 @@ absl::StatusOr<bool> FuseBwdMultiHeadedAttentionBlock(
...
   HloInstruction* fwd_act;
-  if (fwd_config.is_flash_attention()) {
-    auto fwd_act_index = 2;
-    fwd_act = comp->AddInstruction(HloInstruction::CreateGetTupleElement(
-        fwd_fmha_call->shape().tuple_shapes(fwd_act_index), fwd_fmha_call,
-        fwd_act_index));
-  } else {
-    fwd_act = lhs_bmm2_grad_gemm1;
-  }
+  auto fwd_act_index = 2;
```
Comment (Tixxx) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:1396]: maybe just in64_t?

File: xla/service/gpu/cublas_cudnn.h [hunk 4]
```diff
@@ -181,37 +181,22 @@ bool IsCustomCallToDnnNorm(const HloInstruction& hlo);
 
 // The fused_mha_rewriter phase where each of the MHA signatures are pattern
 // matched and rewritten into a custom-call with specific custom-call target.
-// The custom-call target specifies the MHA signature. For example,  BMM1 - Bias
-// - Scale - Mask - Softmax - BMM2 pattern can have the target as
-// cudnn$fmhaBiasScaleMaskSoftmax.
+// The custom-call target specifies the MHA signature. For example,  BMM1 -
+// Scale
```
Comment (akuegel) [inline xla/service/gpu/cublas_cudnn.h:185]: nit: please reflow the comment such that the following line is moved to this line (until 80 character limit)

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc [hunk 5]
```diff
@@ -1113,19 +937,18 @@ MatchBwdResult MatchBwdBmmSoftmaxDropoutBmm(MatchBwdResult previous_result,
...
   // try to pattern match dbias
-  HloInstruction* dS = has_mask ? d_mask : d_softmax;
+  HloInstruction* dS = d_softmax;
   if (dS->users()[0]->opcode() == HloOpcode::kConvert) {
     dS = dS->users()[0];
   }
   if (has_scale) {
-    // bmm1-(scale)-(bias)-(mask)-softmax pattern
-    // users could be dbias besides mask bwd or scale bwd
+    // bmm1-(scale)-(bias)-softmax pattern
```
Comment (akuegel) [inline xla/service/gpu/cudnn_fused_mha_rewriter.cc:945]: nit: I guess this comment can fit into one line? or at least more of the next line can fit in this line

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 2 above)
Comment (Cjkkkk) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:1429]: is_causal_mask is used to indicated mask generation inside cuDNN. If the bias is causal mask, we don't take in bias, we just generate causal mask directly.

File: xla/service/gpu/ir_emitter_unnested.cc (hunk 1 above)
Comment (Cjkkkk) [reply xla/service/gpu/ir_emitter_unnested.cc:1083]: Thanks, I forgot to remove this.

File: xla/service/gpu/cublas_cudnn.h (hunk 4 above)
Comment (Cjkkkk) [reply xla/service/gpu/cublas_cudnn.h:185]: fixed

File: xla/service/gpu/cudnn_fused_mha_rewriter.cc (hunk 2 above)
Comment (Tixxx) [reply xla/service/gpu/cudnn_fused_mha_rewriter.cc:1429]: ah right, thanks.

**Review by Tixxx — APPROVED** [review]

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 6]
```diff
@@ -164,135 +150,48 @@ class CudnnFusedMhaRewriterPipelineTest
...
-  const HloInstruction* fmha;
-
-  SCOPED_TRACE(m->ToString());
-  EXPECT_THAT(
-      m->entry_computation()->root_instruction(),
-      GmockMatch(m::GetTupleElement(
-                     m::CustomCall(&fmha, {kCudnnfMHABmmBmmCallTarget}), 0)
-                     .WithShape(BF16, {16, 16, 256, 64})));
-  TF_ASSERT_OK_AND_ASSIGN(auto gpu_config,
-                          fmha->backend_config<GpuBackendConfig>());
```
Comment (sgerrard) [inline xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:194]: The white spaces here are blocking the merge. The exact error is "Lines should be <= 80 columns wide  [whitespace/line_length] [2]". Looks like too many whitespaces, please fix.

File: xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 7]
```diff
@@ -164,135 +150,48 @@ class CudnnFusedMhaRewriterPipelineTest
...
-  const CudnnfMHABackendConfig& config = gpu_config.cudnn_fmha_backend_config();
-  EXPECT_EQ(config.fmha_scale(), 1.0);
-  EXPECT_EQ(config.dropout_rate(), 0.0);
+region_0.7 {
+  Arg_0.8 = bf16[] parameter(0)
+  Arg_1.9 = bf16[] parameter(1)
+  ROOT maximum = bf16[] maximum(Arg_0.8, Arg_1.9)
 }
 
-TEST_F(CudnnFusedMhaRewriterPipelineTest, BF16Bmm1Bmm2UncanonicalizedPattern) {
```
Comment (sgerrard) [inline xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:255]: The white spaces here are blocking the merge. The exact error is "Lines should be <= 80 columns wide [whitespace/line_length] [2]". Looks like too many whitespaces, or a trailing white space please fix.

File: xla/service/gpu/cudnn_fused_mha_rewriter_test.cc [hunk 8]
```diff
@@ -312,50 +211,48 @@ TEST_F(CudnnFusedMhaRewriterTestHloTest,
             2);
 }
 
-TEST_F(CudnnFusedMhaRewriterPipelineTest,
-       BF16Bmm1Bmm2Pattern_bmm1_rhs_contracting_dim_not_most_minor) {
-  if (skip_reason_) GTEST_SKIP() << *skip_reason_;
-  TF_ASSERT_OK_AND_ASSIGN(
```
Comment (sgerrard) [inline xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:318]: Same error, "Lines should be <= 80 columns wide [whitespace/line_length] [2]", trailing whitespace again?

**Review by sgerrard — CHANGES_REQUESTED** [review]
Please fix blocking style errors.

**Comment by Cjkkkk** [comment]
@sgerrard Why are the errors on the deleted lines?

**Comment by sgerrard** [comment]
The lines that have issues in cudnn_fused_mha_rewriter_test.cc are 194, 255 and 318. Looks like I did mark the old line numbers instead of new.

**Comment by Cjkkkk** [comment]
> The lines that have issues in cudnn_fused_mha_rewriter_test.cc are 194, 255 and 318. Looks like I did mark the old line numbers instead of new.

I renamed the problematic string to a shorter name.

**Review by akuegel — APPROVED** [review]

**Comment by mooskagh** [comment]
@Cjkkkk this change happens to cause a performance regression for this model:
[regression-11717.hlo.txt](https://github.com/openxla/xla/files/15146913/regression-11717.hlo.txt)

The runtime (on H100) goes from 40ms, to 6200ms.  The GPU part seems to be faster, but before every flash attention block (which this network has 14), there is a call to `CudnnGraph::Prepare()` that takes a half of a second.

**Pull request closed without merging**